### ISSUE TITLE
[Contribution] qvm-screenshot-tool

### ISSUE THREAD

**Issue by marmarek** [QubesOS/qubes-issues#953]
**Previous community devs:**  @evadogstar, @ben-grande
**Last active dev repo:** https://github.com/ben-grande/qubes-qvm-screenshot-tool

-----

**Reported by axon on 16 Feb 2015 11:02 UTC**
The ability send dom0 screenshots directly to an AppVM/DispVM from the screenshot app is an oft-requested feature (see below). The ability to transfer saved screenshot files from dom0 to other VMs is [ already available](CopyToDomZero), but it is neither obvious nor easy for most users to do this.

User requests/queries about this:
https://groups.google.com/d/topic/qubes-devel/m8TfyvSqvf4/discussion
https://groups.google.com/d/topic/qubes-devel/CwSPqtPYTRQ/discussion
https://groups.google.com/d/topic/qubes-devel/_a7KxHbkSJo/discussion
https://groups.google.com/d/topic/qubes-users/l6vOqhsd7ss/discussion
https://groups.google.com/d/topic/qubes-users/etxwrc6rsIM/discussion
https://groups.google.com/d/topic/qubes-users/_7FzKv6eJqA/discussion

Migrated-From: https://wiki.qubes-os.org/ticket/953

**Comment by marmarek** [comment]
Some details here: https://groups.google.com/d/msgid/qubes-devel/20150708022217.GH900%40mail-itl

**Comment by ptitdoc** [comment]
What about a tool in dom0 that scan changes in ~/QubesOutgoing/[VMname]/ folder and send it to the AppVM ~/QubesIncoming/dom0 when available ?

For instance, when you take screenshots, you just have to save it to the right folder.

**Comment by marmarek** [comment]
I think the best solution is to use #1324 (qvm-copy-to-vm tool for dom0)
and make it a handler for screenshooting tool.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by mfc** [comment]
fyi with KSnapshot (the current screenshot tool) one cannot screenshot a specific window, freehand region, etc., even when selected. Such functionality would be quite helpful for creating Qubes documentation.

**Comment by marmarek** [comment]
Are you sure? I've just tried and it seems to work well:

1. Press Print Screen to start the tool
2. Select capture mode
3. Click "Take a New Snapshot"
4. Select region/window/whatever
5. Confirm with Enter (or double-click) - instructions are on the
screen.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by mfc** [comment]
never mind! worked great. 

I misunderstood how it worked, I thought changing the settings would change the default of pressing Print Screen. Instead one has to select Take a New Snapshot.

**Comment by Jeeppler** [comment]
Ksnapshot is getting a new version for KDE 5 based on [KScreenGenie](https://blogs.kde.org/2015/08/12/ksnapshot-next). On which KDE version will Qubes R3.2 or 4.0 be based on?

**Comment by andrewdavidwong** [comment]
> On which KDE version will Qubes R3.2 or 4.0 be based on?

The plan is to use KDE 5.

**Comment by Jeeppler** [comment]
Is there any usable Qubes R4.0 or R3.2 available to develop such a feature?

**Comment by marmarek** [comment]
I'm going to upload some preliminary R3.2 image somehow today/tomorrow.
Anyway this isn't much different than R3.1 - see: https://github.com/QubesOS/qubes-issues/issues/1324#issuecomment-155655636

**Comment by Jeeppler** [comment]
@marmarek I think the Desktop Entry solution, you mentioned in the comment, has several draw backs. First it does not work on all desktop environments and secondly you have to open the PNG file, as far as I understand it. 

For my understanding it would be better to have a screen shoot tool which invokes the `qvm-move-to-vm` command. The tools should move the screenshot directly to a qube and only ask to start the destination qube if the destination qube is not running.

I think KSnapshot is a really nice tool  to take snapshots with, but I don't know if it works reliable on XCFE or other desktop environments.

**Comment by marmarek** [comment]
It is simply available as one of applications to open screenshot with, see here:
![screenshot-save](https://cloud.githubusercontent.com/assets/726704/15275753/6c6f41f2-1ad4-11e6-9def-875455b5f132.png)

This one is from Xfce4. In KDE it lands under "Open with" menu (or "Send to" in KDE4).

**Comment by Jeeppler** [comment]
Looks good, but what should be developed or created to be able to close this issue?

**Comment by marmarek** [comment]
A tool to ask for target VM name - `qvm-move-to-vm-prompt`

**Comment by Jeeppler** [comment]
What would be the preferred programming language and graphical toolkit to develop such a prompt? Should it look like the prompts Qubes OS already has to copy files from one VM to another?

**Comment by marmarek** [comment]
As usual - python. As for toolkit - currently it is mostly in Qt (especially the current qrexec confirmations). But since the long term goal is to implement new Qubes Manager in GTK and to add GNOME support, GTK is also ok. If it doesn't matter for you, I'd choose GTK.

**Comment by Jeeppler** [comment]
Is it possible to have access to the Qubes API via Python3?

**Comment by marmarek** [comment]
On Thu, May 19, 2016 at 01:57:54AM -0700, Jeppler wrote:
> Is it possible to have access to the Qubes API via Python3?

Not yet, unfortunately...

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by greenrd** [comment]
Unfortunately "send to" in KSnapshot does not currently work - it says "qvm-run: error: To many arguments" [sic]. Should I file another bug for that?

**Comment by marmarek** [comment]
@greenrd it isn't done yet - "Other application" option you see there is not what is meant to be used here.

**Comment by Jeeppler** [comment]
I think it would be a good idea to schedule this task till Qubes OS uses KDE 5 and especially Python 3.

**Comment by dumbl3d0re** [comment]
Jeppler:
> I think it would be a good idea to schedule this task till Qubes OS uses KDE 5 and especially Python 3.

I would like to expand on this as propose an application for video
capturing. This for future educational material for new users, and video
is easier to digest than long pages of text with screenshots.

**Comment by Jeeppler** [comment]
@dumbl3d0re your idea should be in a new issue. In general the idea of having a video capture tool in Qubes OS is a good idea. 

This issue is primarily talking about using KSnapshots 'send to' function. The question is which video capturing tool does KDE or linux in general provide for capturing videos?

**Comment by andrewdavidwong** [comment]
[On 2016-07-02 13:57, Eva Star wrote:](https://groups.google.com/d/msgid/qubes-users/cfce52ee-92d3-458c-8b67-a1a7f4cf0d3e%40googlegroups.com)
> Okey. I released tool that can automatically capture 
> fullscreen/windows/regions and upload it to AppVM and imgurl automaticaly.
> 
> Now it is in beta state and tested only with Qubes 3.2rc1, but I think it 
> will work with other Qubes starting from R3.0 or R3.1 (where qvm-run is 
> available)
> 
> Full description and download you can find here: 
> 
> https://github.com/evadogstar/qvm-screenshot-tool
>  
> Test it and I hope you will be happy with it as I am :) Because screenshots 
> is one weak side of Qubes (before this tool done ;) 
> 
> Plans: 
> * add editor for images at AppVM to blur some arias on image... (suggest 
> are welcome)
> * multiple selections -> one screenshots -> upload it
> * delayed screenshots
> * maybe uploading any existed image from dom0 to imgurl after selecting it 
> on dialog
> 
> Enjoy but remember that is only beta
>

**Comment by evadogstar** [comment]
qvm-screenshot-tool updates and now  support editing of image at dom0 before uploading it to AppVM / Imgurl 

https://i.imgur.com/BZHFvJq.png

![2016-07-04-204252](https://cloud.githubusercontent.com/assets/20259395/16567521/54046806-4228-11e6-96e6-450ec3885475.png)

**Comment by evadogstar** [comment]
@marmarek What is the right way to store some temp.config values? Create some file with value or keys? Or use something other? Thanks

**Comment by marmarek** [comment]
It depends on exact purpose and use case, but temp file seems to be a good solution.

**Comment by marmarek** [comment]
I'm not sure if I understand your question correctly, can you elaborate? @evadogstar

**Comment by evadogstar** [comment]
@marmarek 
1) Now, I want to store setting: i.e. last "VM name" from where image was uploaded to imgurl service. This need to reopen last closed dialog by simply choose "Open last" from dom0 menu.  
Estimated solution: file "~/.config/qvmscreenshot/lastvm.cfg" with content "personal" (to read it later)

2) Maybe, I want to implement some log of all uploaded urls. It's will give users ability to review. 
Estimated solution: file "~/.config/qvmscreenshot/log.txt" with full log.

3) Config for script settings. Now, settings stored at the source. How to store/read them from separate file/QubesDB? 

I do not want to reinvent the wheel with config)

**Comment by marmarek** [comment]
> 1) I want to store setting: i.e. last "VM name" from where image was uploaded to imgurl service. This need to reopen last closed dialog with results again by simply choose "Open last" from dom0 menu. I found that this will be useful feature.
> Estimated solution: file "~/.config/qvmscreenshot/lastvm.cfg" with content "personal" (to read it later)

Yes, this seems like a good idea. Take a look [here](http://stackoverflow.com/a/16260438) for more robust config handling from scripts (have one file, instead of one-per-value)

> 2) Maybe, I want to implement some log of all uploaded urls.
> Estimated solution: file "~/.config/qvmscreenshot/log.txt" with full log.

I think logs belong to `~/.local/share/...`, at least X server store its log in `~/.local/share/xorg/Xorg.0.log`.

> 3) "Config" for some script settings (available to change by user). Now, settings stored at the source. How to store/read them from separate file/QubesDB? What is better?

See "1". Maybe have two files - one with user changeable settings, the other with state from previous run?

**Comment by evadogstar** [comment]
@marmarek About screencasts. All of such tools have a lot of dependencies (i.e. ffmpeg and other). 
Is it possible/reasonable to add them to dom0? What you can say about screencasting ?
I found good tool for screencasting:
https://github.com/colinkeenan/silentcast

**Comment by v6ak** [comment]
In my opinion, there are generally two concerns when you are considering adding a 3rd party SW to dom0:

1. Do I trust it? (Do I trust there are no intentional backdoors?)
2. Is it secure enough?

The first one is rather a matter of choice than a technical question.

The security concern is the interesting one. While level of acceptable risk is not a technical question, the risk itself (at least partially) is. I believe there are not-yet-known vulnerabilities related to complex formats processing in ffmpeg and some other software you are considering. Exposing dom0 to those vulnerabilities can be considered at least strongly against Qubes philosophy. (If it is acceptable or not, it is up to you.) However, the key question is: Are those potential vulnerabilities likely to be exploitable? I don't think so. I expect the screenshots (in a simple bitmap format) and potentially a sound (in similarly simple raw format) to be the only untrusted inputs.

In short, after rather brief review, it looks OK to me provided that you trust the software not to be intentionally malicious.

**Comment by marmarek** [comment]
Generally, if input is trusted, even complex tools are not a big problem.
The question is whether screen content (as a bitmap) can be treated as
trusted input. I find it very unlikely to crash/exploit some image/video
processor by just a bitmap (without control over its metadata). So,
should be safe.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by andrewdavidwong** [comment]
@marmarek: Is this feature slated for inclusion in 4.0?

**Comment by evadogstar** [comment]
@andrewdavidwong my tool work well and do more than requested at first post

**Comment by ptitdoc** [comment]
Is there any work in progress on this topic, or can we start contributing ?

If I followed the discussion properly there are two approaches:
- @evadogstar already wrote a screenshot client that allows editing screenshots, sending it to specific VM, eventually edit the screenshot (in dom0?) and keep the history of where the screenshot is being sent.
  - PROs are features
  - CONs are dependencies, specific code for file upload, and edition needs to be added to dom0, screenshot feature is not properly integrated into the panel (except if we create a specific action button into the panel)

- @marmarek is proposing to integrate into existing screenshot tools by using qvm-copy-to-vm the same way you can open a file with qvm-open-in-dvm in an AppVM, but calling a wrapper script. This requires writing a .desktop file and a script that wraps qvm-copy-to-vm with a widget that asks which AppVM the screenshot will be sent to.
  - PROs are flexibility of this approach, and the inocuity of code that needs to be written for dom0.
  - CONs are missing features such as screenshot edition (which could be delivered directly by the screen shot tool), and file upload to specific web services, and code that is still missing (however the script qvm-mru-entry available in AppVMs can probably be used).

To summarize, the second solution allows keeping dom0 clean but is missing  automated features requested by end-users such as "share my screenshot" or "annote my screenshot". This would require writing specific Qubes RPCs to "do something with my data coming from dom0 inside my AppVM", and I fear that it could cause incidents such as "oops I uploaded sensitive dom0 data by error".

**Comment by ptitdoc** [comment]
I can confirm that qvm-mru-entry works in dom0 without any change.

**Comment by ptitdoc** [comment]
@marmarek: can you provide your desktop entry somewhere ? I did not managed to create a desktop file that works properly with xfce4-screenshot.

**Comment by marmarek** [comment]
Sure, it looks like this:
```
[Desktop Entry]
Type=Application
Name=Save image to testvm
Exec=qvm-move-to-vm testvm %F
MimeType=image/png;
NoDisplay=true
```

**Comment by ptitdoc** [comment]
After importing qvm-mru-entry in dom0 (in /usr/bin), the following two files are sufficient to make this working:

```
/usr/share/applications/qubes-movetovm.desktop:
[Desktop Entry]
Name=SaveToAppVM
GenericName=Save image to AppVM
Comment=Open an AppVM selector and save image in the selected AppVM
Exec=/usr/lib/qubes/qvm-move-to-vm.gnome
Icon=qubes-appmenu-select
Type=Application
Terminal=false
NoDisplay=true
Categories=Graphics;Viewer;
MimeType=application/x-navi-animation;image/bmp;image/x-bmp;image/x-MS-bmp;image/gif;image/x-icon;image/jpeg;image/png;image/x-portable-anymap;image/x-portable-bitmap;image/x-portable-graymap;image/x-portable-pixmap;image/x-cmu-raster;image/x-sun-raster;image/x-tga;image/tiff;image/vnd.wap.wbmp;image/x-xbitmap;image/x-xpixmap;image/svg;image/svg+xml;image/x-png;image/xpm;image/x-ico;image/x-pcx;
X-Desktop-File-Install-Version=0.22
```
```
/usr/lib/qubes/qvm-move-to-vm.gnome:
#!/bin/sh
VM=$(qvm-mru-entry --title="File Copy" --text="Enter the destination domain name:" --mrufile "qvm-mru-filecopy")
if [ X$VM = X ] ; then exit 0 ; fi

qvm-move-to-vm $VM "$@"
```

There are of course room for improvement such as:
 - providing a drop-down list in qvm-mru-entry, but as qvm-mru-entry remembers the last entered vm, it is not a big improvement.
 - providing some feedback to say that the file has been copied successfully, but again, I thing there is a popup if the copy failed, and a success message is just another spam popup for me.

**Comment by evadogstar** [comment]
@ptitdoc It's not a lot of dependencies, because almost all tools already preinstalleted at the dom0. What do you mean with "integration into the panel" ? I use my tool every day and do not know why it's need any integration to the panel, because it's very simple to use key binding to run the script and it's traditional way to make screenshots and manipulate them. 
About "special code to upload screenshots". It can be  pre-installeted at the template, then the script will use available code from selected appvm. I do not separate the code now, because it's a lot of additional actions to add 'upload script' to every template manually if the script can do it for you from dom0 for any unix-based appvm after selecting it from the list.

**Comment by jpouellet** [comment]
There seems to be some resistance to the upstreaming of @evadogstar's implementation due to its size and the question of whether editing screenshots should be done in dom0 in the first place.

In the interest of having something useful available, I've made a minimal alternative. It's quite simple, and perhaps best explained by screenshots ;)

You press the Print Screen key (or in some way invoke xfce4-screenshooter) and get SaveImageInVM as an option:
![screenshot_2017-06-27_04-18-58](https://user-images.githubusercontent.com/1231207/27577831-173d47bc-5af0-11e7-858c-3254fb20c3eb.png)

Then you get a list of running VMs, and select where to save the image:
![screenshot_2017-06-27_04-19-58](https://user-images.githubusercontent.com/1231207/27577876-395632e6-5af0-11e7-81ca-71313efd793f.png)

And then it saves the screenshot there and opens the location where it saved it (~/QubesIncoming/dom0) in a file manager:
![screenshot_2017-06-27_04-21-03](https://user-images.githubusercontent.com/1231207/27577934-69c50d62-5af0-11e7-8505-85fe0856ae8b.png)
so you can then do whatever with it (like open it in gimp or whatever you prefer in an AppVM, instead of solidifying a dependency on ImageMagick in dom0 and asking people to actually learn to use the (IMO ugly and unfriendly) ImageMagick GUI)

Source here: https://github.com/jpouellet/qubes-screenshot-helper (qubes-builder friendly)

I'm happy to change the name of the repo or propose this as a PR against qubes-desktop-linux-common if preferred.

It could of course be improved, but lets start somewhere. This is small and probably does all of what most users really need. Right now everybody needs to `qvm-copy-to-vm` their screenshots from the command line, and that's not a friendly workflow.

**Comment by jpouellet** [comment]
@evadogstar I don't mean to detract from your work. You've made a useful thing, and I thank you for contributing. I'm just trying to move things along, and I also prefer to do as little as we can in dom0.

**Comment by marmarek** [comment]
This, and others, is blocked on setting up @Qubesos-contrib finally. To be honest, most of technical stuff is done - there are even [package repository definitions](https://github.com/QubesOS/qubes-meta-packages/tree/master/repos). What is missing is a procedure to handle it:
 - how to request package being added there
 - what are inclusion criteria
 - how to handle updates
etc

Some of this was already discussed on qubes-devel, but not all. And we need to clarify all that and document it somewhere. @andrewdavidwong do you want to draft it? Anyway probably worth a separate ticket.

**Comment by andrewdavidwong** [comment]
@marmarek: Will do.

**Comment by evadogstar** [comment]
@jpouellet No problem, but with my tool it's also possible to easy save&move screenshot to AppVM then edit it with GIMP. ImageMagick is already there at dom0, but I'm also think that it's very ugly. If we will have some other pre-installeted image editor at the fedora template, then we can open screenshot with it, edit, then upload (or what user want).

P.S. Personally, I do NOT like the way to use default Qubes screenshot tool and "Open With" menu from it, because this tool can be changed (and it already changed on my mind from kde screenshot tool to current one. And it will changed on the future) and "Open With" will totally broken, while my implementation will continue working.

**Comment by jpouellet** [comment]
This came up on the mailing lists again: https://groups.google.com/d/topic/qubes-users/q1fizu46Pbs/discussion

What's this still blocking on?

If it's still waiting on @QubesOS-contrib, can we at least upstream at least my [minimalist implementation](https://github.com/jpouellet/qubes-screenshot-helper) in the mean time? It might not be the long-term ideal solution, but it provides a usable solution today to a real user issue which keeps coming back up.

**Comment by jpouellet** [comment]
I've updated [my minimal screenshot tool](https://github.com/jpouellet/qubes-screenshot-helper) for R4. Any chance at getting it (or anything equivalent) merged? Users *still* have no easy way to get screenshots out of dom0.

**Comment by unman** [comment]
On Thu, Mar 22, 2018 at 04:40:01PM -0700, Jean-Philippe Ouellet wrote:
> I've updated [my minimal screenshot tool](https://github.com/jpouellet/qubes-screenshot-helper) for R4. Any chance at getting it (or anything equivalent) merged? Users *still* have no easy way to get screenshots out of dom0.
> 

I use KSnapshot.
What's wrong with qvm-copy-to-vm run in dom0?
Easy enough

**Comment by jpouellet** [comment]
Easy enough for you and me, sure, but not everyone. I only developed the tool I did because I was asked to by someone for whom the command line was not a good solution.

Given that questions about getting screenshots out of dom0 keep coming up ([[1]](https://groups.google.com/forum/#!topic/qubes-users/l6vOqhsd7ss) [[2]](https://www.reddit.com/r/Qubes/comments/5l0th4/screenshots/) [[3]](https://groups.google.com/forum/#!topic/qubes-users/QvabfH2GjB4) [[4]](https://groups.google.com/forum/#!topic/qubes-users/q1fizu46Pbs/discussion) & more that I don't feel like searching for), and how many people have independently rolled their own screenshot scripts at this point ([[1]](https://jrruethe.github.io/blog/2015/09/17/screenshots-in-qubes/) [[2]](https://github.com/evadogstar/qvm-screenshot-tool) [[3]](https://github.com/jpouellet/qubes-screenshot-helper) [[4]](https://groups.google.com/forum/#!topic/qubes-users/ruGz2c3qIJ8) [[5]](https://groups.google.com/d/topic/qubes-devel/yevq_UDM84s/discussion) [[6]](https://github.com/Aekez/Qubes-Scripts/wiki/Streamlining-the-Screenshotting), etc.), I think the case for having something easier to use out of the box has already been well proven.

**Comment by andrewdavidwong** [comment]
Pinging @marmarek

**Comment by jpouellet** [comment]
I'm not lobbying for my particular implementation here, I just want users to have something that works for them.

Personally, I think the criteria here should be:
- easy to use
- no additional attack surface
- no new dependencies
- supports R3.2 and R4

I believe mine fits all of the above, and it's already a component you can just drop into qubes-builder, but anything else is welcome too.

I don't like the way mine chooses the destination VM, and I think it'd be nice to use have it do a qrexec call instead and re-use the nice dom0 qrexec target selection GUI, which would also give the user the flexibility to use `,target=some-screenshot-recipient-vm` in policy, or override the hypothetical `qubes.ReceiveScreenshot` implementation on a per-VM basis, but I didn't find an easy way to force qrexec calls originating from dom0 to observe `ask` behavior, rather they are automatically approved for any destination vm.

Also, I think [799] had a good idea [here](https://groups.google.com/d/topic/qubes-users/ruGz2c3qIJ8/discussion) with copying the screenshot into the clipboard of the frontmost VM, but that introduces a VM-side dependency on xclip. Honestly I think adding xclip to templates by default is not a bad idea (and I invariably end up installing an equivalent anyway) but idk if that's just a personal preference or if it's actually justified.

**Comment by mike4420** [comment]
I use i3 and wrote this simple script I invoke with a shortcut after taking a screenshot with xfce4-screenshooter.
It simply copies all files newer than 5 minutes from dom0:~/Pictures to the current qube.
Gets job done even though it's dumb simple. Don't know how it can be done in other WMs and/or if it makes sense there, but it really fits well into i3 flow.

```
#!/bin/bash

. /usr/local/bin/qubes-inc

main() {
    local vm=$(get_vm)
    if [[ -n "$vm" ]]; then
        find ~/Pictures -type f -maxdepth 1 -mmin -5 | xargs qvm-copy-screenshot "$vm"
    fi
}

main
```

get_vm is a function defined in qubes-inc returning the VM which the current window is running on.
qvm-copy just runs qvm-copy-to-vm for a given file from ~/Pictures and then moves that file to ~/Pictures/copied/.

**Comment by mryellow** [comment]
Thanks for the work formalising something @jpouellet. Works well. 

For some reason the approach from @marmarek of `Exec=qvm-move-to-vm testvm %F` didn't work. Couldn't find a log to help debug exactly why.

https://github.com/QubesOS/qubes-issues/issues/953#issuecomment-286129671

One minor suggestion would be to use the envvar `$USER` here just in-case they've changed it from default:

`qvm-run -p "$vm" 'xdg-open /home/$USER/QubesIncoming/dom0' </dev/null`

**Comment by Thovthe** [comment]
What's the status of this?  Are any of these solutions included in the repos yet?  Is the contributor's repo set up yet?

**Comment by andrewdavidwong** [comment]
> What's the status of this? Are any of these solutions included in the repos yet? Is the contributor's repo set up yet?

I believe the current status is for @marmarek to review the code contributed by @evadogstar in https://github.com/evadogstar/qvm-screenshot-tool.

**Comment by ninavizz** [comment]
Pinging @marmarta... any chance I could see how this is currently invoked by the user, and how it works? Would there be time for me to make any text, icon, or interaction reccs/improvments, if I have any?

**Comment by ninavizz** [comment]
^ ...and only pinging Marta, because Marek is already spread too thin... but if Marek is the better person for this, then I'll gladly work with Mr Marek on this!

**Comment by marmarek** [comment]
See screenshots on https://github.com/evadogstar/qvm-screenshot-tool
The plan is to connect it to PrintScr button as a default action, instead of the current xfce4-screenshooter.
The UI for making screenshots is not that nice as in xfce4-screenshooter, but on the other hand, it allows to copy the image out of dom0 to various places (selected vm, imgur upload etc) out of the box.

**Comment by ninavizz** [comment]
Yeah, not having to use the CLI is a pretty big deal. I also really love that it allows for annotations.

I previously saw those, and they're rad—but they lack the context to let me know what was invoked to show which, and how it all works. If that makes sense. Full screenshots are more helpful than the tidy little windows that are in that repo.

Below is kinda where I'm at with noodling on Tray icon/functionality stuff. It feels like this would be the best way imho to invoke the screenshot tool (from the tray vs from a menu)—but I don't know what the invoking would prompt. I also don't feel that Tray icons need to invoke dropdown menus—if anything I feel it'd be more helpful if they did not longer-term. But the context needs to suit those usability things. 

Thoughts?

![image](https://user-images.githubusercontent.com/8262612/75198208-e7e92480-5714-11ea-8775-a0ef2a47bc36.png)

**Comment by marmarta** [comment]
In the future we will be moving more and more towards removing things from dom0 - some tools will be in GUI VM (like most widgets, including things like batter widgets), some tools in Audio VM (like volume controls). I wonder if we should make such things also distinguished visually, or just consider those "administrative VMs" and not color their widgets.

**Comment by v6ak** [comment]
Maybe it depends:

* The case of Audio VM is close to sys-net. After all, a headset it can be even connected wirelessly, for example through Bluetooth. Moreover, this seems to be the purpose of having a separate VM.
* GUI VM is quite special case. It draws all the windows and the border. If we decided to have a coloring for the GUI VM, the VM would be responsible of its own coloring… So, in the case of GUIVM, the coloring does not directly serve for any security purpose. However, I am still in favor of adding at least a label in the title, so the user is informed about the compartmentalization.

**Comment by evadogstar** [comment]
Will my tool work with GUI domain the same way?

**Comment by marmarek** [comment]
I haven't tried yet, but probably some changes will be needed. One difference may be some qrexec confirmation prompts and/or required policy changes (when calling qvm-run from dom0, you don't have prompts).
To avoid this, it may be required to implement some of the actions (like the upload helper) as a dedicated qrexec service, and setup policy appropriately. Also, copying image should be done with qvm-copy-to-vm, not qvm-run.

**Comment by fepitre** [comment]
`scrot` is a dead package since Fedora 32. We cannot install/use it under latest R4.1.

**Comment by tlaurion** [comment]
@fepitre @marmarek 
```
[user@dom0 ~]$ sudo qubes-dom0-update qubes-repo-contrib
[user@dom0 ~]$ sudo qubes-dom0-update qvm-screenshot-tools
```

No packages avail. Testing because not moved over to stable.

```
[user@dom0 ~]$ sudo qubes-dom0-update --enablerepo=qubes-contrib-dom0-r4.0-current-testing qvm-screenshot-tool
Using sys-whonix as UpdateVM to download updates for Dom0; this may take some time...
--> Running transaction check
---> Package qvm-screenshot-tool.x86_64 0:0.7-0.fc25 will be installed
--> Processing Dependency: scrot for package: qvm-screenshot-tool-0.7-0.fc25.x86_64
--> Running transaction check
---> Package scrot.x86_64 0:0.8-14.fc24 will be installed
--> Processing Dependency: libImlib2.so.1()(64bit) for package: scrot-0.8-14.fc24.x86_64
--> Processing Dependency: libgiblib.so.1()(64bit) for package: scrot-0.8-14.fc24.x86_64
--> Running transaction check
---> Package giblib.x86_64 0:1.2.4-24.fc24 will be installed
---> Package imlib2.x86_64 0:1.4.9-1.fc25 will be installed
--> Finished Dependency Resolution
qvm-screenshot-tool-0.7-0.fc25.x86_64.rpm                   |  18 kB  00:00     
/var/lib/qubes/dom0-updates/packages/imlib2-1.4.9-1.fc25.x86_64.rpm already exists and appears to be complete
/var/lib/qubes/dom0-updates/packages/giblib-1.2.4-24.fc24.x86_64.rpm already exists and appears to be complete
/var/lib/qubes/dom0-updates/packages/scrot-0.8-14.fc24.x86_64.rpm already exists and appears to be complete
Successfully verified /var/lib/qubes/dom0-updates/packages/giblib-1.2.4-24.fc24.x86_64.rpm
Successfully verified /var/lib/qubes/dom0-updates/packages/imlib2-1.4.9-1.fc25.x86_64.rpm
Successfully verified /var/lib/qubes/dom0-updates/packages/qubes-repo-contrib-4.0.6-1.fc25.noarch.rpm
ERROR: could not verify /var/lib/qubes/dom0-updates/packages/qvm-screenshot-tool-0.7-0.fc25.x86_64.rpm
Successfully verified /var/lib/qubes/dom0-updates/packages/scrot-0.8-14.fc24.x86_64.rpm
ERROR: could not verify one or more packages
```

**Comment by marmarek** [comment]
Maybe contrib key is not imported? Try `sudo rpm --import /etc/pki/rpm-gpg/RPM-GPG-KEY-qubes-4-contrib-fedora`

**Comment by tlaurion** [comment]
@marmarek 

```
[user@dom0 ~]$ sudo rpm --import /etc/pki/rpm-gpg/RPM-GPG-KEY-qubes-4-contrib-fedora 
[user@dom0 ~]$ sudo qubes-dom0-update --enablerepo=qubes-contrib-dom0-r4.0-current-testing qvm-screenshot-tool
Using sys-whonix as UpdateVM to download updates for Dom0; this may take some time...
--> Running transaction check
---> Package qvm-screenshot-tool.x86_64 0:0.7-0.fc25 will be installed
--> Processing Dependency: scrot for package: qvm-screenshot-tool-0.7-0.fc25.x86_64
--> Running transaction check
---> Package scrot.x86_64 0:0.8-14.fc24 will be installed
--> Processing Dependency: libImlib2.so.1()(64bit) for package: scrot-0.8-14.fc24.x86_64
--> Processing Dependency: libgiblib.so.1()(64bit) for package: scrot-0.8-14.fc24.x86_64
--> Running transaction check
---> Package giblib.x86_64 0:1.2.4-24.fc24 will be installed
---> Package imlib2.x86_64 0:1.4.9-1.fc25 will be installed
--> Finished Dependency Resolution
qvm-screenshot-tool-0.7-0.fc25.x86_64.rpm                   |  18 kB  00:00     
/var/lib/qubes/dom0-updates/packages/imlib2-1.4.9-1.fc25.x86_64.rpm already exists and appears to be complete
/var/lib/qubes/dom0-updates/packages/giblib-1.2.4-24.fc24.x86_64.rpm already exists and appears to be complete
/var/lib/qubes/dom0-updates/packages/scrot-0.8-14.fc24.x86_64.rpm already exists and appears to be complete
Successfully verified /var/lib/qubes/dom0-updates/packages/giblib-1.2.4-24.fc24.x86_64.rpm
Successfully verified /var/lib/qubes/dom0-updates/packages/imlib2-1.4.9-1.fc25.x86_64.rpm
Successfully verified /var/lib/qubes/dom0-updates/packages/qubes-repo-contrib-4.0.6-1.fc25.noarch.rpm
ERROR: could not verify /var/lib/qubes/dom0-updates/packages/qvm-screenshot-tool-0.7-0.fc25.x86_64.rpm
Successfully verified /var/lib/qubes/dom0-updates/packages/scrot-0.8-14.fc24.x86_64.rpm
ERROR: could not verify one or more packages

```

**Comment by TheOne-Z** [comment]
@marmarek @tlaurion 
I replicated this and it seems that you cannot verify any packages from `qubes-cotrib-repo` either.

**Comment by andrewdavidwong** [comment]
> @marmarek @tlaurion
> I replicated this and it seems that you cannot verify any packages from `qubes-repo-contrib` either, at least in dom0.

This seems like a more general bug, so I've opened #6124 for it.

**Comment by icequbes1** [comment]
> Maybe contrib key is not imported? Try `sudo rpm --import /etc/pki/rpm-gpg/RPM-GPG-KEY-qubes-4-contrib-fedora`

This is essentially the issue - the updatevm doesn't get the new key from `etc/yum.repos.d/qubes-contrib-dom0-r4.0.repo` when running qubes-dom0-update the second time.

Quick hack: kill `/var/lib/qubes/dom0-updates` in the updatevm (sys-whonix) after installing qubes-repo-contrib, then install qvm-screenshot-tool.

e.g.,

```
# add contrib repo
sudo qubes-dom0-update qubes-repo-contrib
# clear updatevm dom0 updates cache
qvm-run sys-whonix 'sudo rm -rf /var/lib/qubes/dom0-updates'
# install contrib packages
sudo qubes-dom0-update --enablerepo=qubes-contrib-dom0-r4.0-current-testing qvm-screenshot-tool
```

**Comment by tlaurion** [comment]
Question for devels. Wouldn't it be possible to change the alt-printscreen and printscreen shortcuts in XFCE to ease UX at installation of package?

**Comment by marmarek** [comment]
Totally makes sense. There is a little technical issue: Xfce copies some of the configuration at first start and at this point there is no longer a place for system-wide default that is easy to override by the user - there is only that user place. So, [setting the default works](https://github.com/QubesOS/qubes-desktop-linux-xfce4/commit/9459331a3772decbaa3948cf6f45d98fd87ec800) only before the very first user login (after initial install). Changing that later requires [quite elaborate hack](https://github.com/QubesOS/qubes-desktop-linux-xfce4/commit/53584fc3968988d4f0c23227c9f5b550928b90d5). As linked - we've done both before, but I'd like some less hacky solution...

**Comment by dylangerdaly** [comment]
What's the latest with this? Scrot isn't available for R4.1

**Comment by dylangerdaly** [comment]
Are we going to maintain scrot, and by extension libgiblib?

**Comment by marmarek** [comment]
See discussion at https://github.com/resurrecting-open-source-projects/scrot/issues/22
I don't think maintaining scrot is the way to go - if the above won't be resolved, I'd rather consider using some other maintained tool instead - there doesn't seem to be much required from it, shouldn't be that hard to switch.

**Comment by dylangerdaly** [comment]
Fair enough, I'll keep an eye out.

**Comment by jpouellet** [comment]
I (and a couple others) still use
https://github.com/jpouellet/qubes-screenshot-helper regularly. It's clean,
minimal, and integrates with the default xfce4-screenshooter tool which is
both still maintained and already included in dom0 by default.

All that should be necessary for adoption is forking the repo under an
official Qubes org (-contrib wasn't ready for use when this was written,
nor do I really think working screenshots should be some awkward 3rd-party
experience for users) and inclusion/installation of the package in the
default installer iso.

It could optionally be improved to use the default qubes-RPC UI (instead of
a Zenity dialog) for selecting the target VM in which to save the
screenshot, but, IIRC this was annoying to implement at the time due to the
qubes-RPC dialog's tight coupling with actual performing of RPC. Maybe it'd
be easier today, I haven't looked. I haven't even thought about this in
quite a while since the tool continues to "Just Work" as-is since I wrote
it.

>

**Comment by marmarek** [comment]
Done, qubes-snapshot-helper is added to standard repos for R4.0 and R4.1 - current-testing for now.

**Comment by dylangerdaly** [comment]
> *qubes-screenshot-helper is added to standard repos for R4.0 and R4.1 - current-testing for now.

**Comment by evadogstar** [comment]
Look like it is possible to change scrot in my tool to maim.
https://github.com/naelstrof/maim

I can do that at next version, but... not sure now.
1) is it available at fedora 33? It is on fedora 32.
2) I can't test it, because I'm on old Qubes...

And there is already 0.8 version on my repo of qvm-screenshot-tool

**Comment by fepitre** [comment]
> Look like it is possible to change scrot in my tool to maim.
> https://github.com/naelstrof/maim
> 
> I can do that at next version, but... not sure now.
> 
>     1. is it available at fedora 33? It is on fedora 32.
> 
>     2. I can't test it, because I'm on old Qubes... and there is no maim in the current repo to test it...
> 
> 
> And there is already 0.8 version on my repo of qvm-screenshot-tool

Don't worry about version. Just do the modification code into a separate branch. Test and validate from your side and open a PR into `https://github.com/QubesOS-contrib/qubes-qvm-screenshot-tool`.

**Comment by evadogstar** [comment]
@fepitre please test it, will works only with maim
https://github.com/evadogstar/qvm-screenshot-tool/blob/Qubes4.1/qvm-screenshot-tool.sh

**Comment by fepitre** [comment]
@evadogstar I've tested to screenshot a portion of window and sent it to appvm. It works. If I understand, `maim` is not available for Fedora 25? If not, I would need to create a branch on `QubesOS-contrib` repository. Do you mind to create a PR please?

**Comment by evadogstar** [comment]
@fepitre About Fedora 25 and maim, look like yes, it's not available...

It is possible to integrate everything (scrot and maim) into one release. But don't think that it's a good idea. Better if previous <R4.0 Qubes users will use current scrot version 0.7/0.8 that works and new users will be on new maim version. Don't think that it request other branch... @marmarek 

p.s. The code of the tool also possible to clear from not needed checks for `maim`, if my tool will be integrated into R4.1. 
I don't know how to create PR. 
Anyway issue almost solved :smile:

**Comment by fepitre** [comment]
I think we are going to do a little wrapper or something to check which one to use. It's better than doing branches and pretty easy even from package point of view. Don't worry about PR, I'll cherry pick your commits and finish the rest.

**Comment by evadogstar** [comment]
@fepitre thanks. When I will be on R4.1 I will look on the code again and maybe add some more features. Looks like maim have some new interesting things. Also we really need some basic editor, because current IM tool that I use to make captions is hard. So, any other light editor at dom0 will be a thing.

**Comment by fepitre** [comment]
Does anyone can test this please: https://github.com/QubesOS-contrib/qubes-qvm-screenshot-tool/pull/1. I've refactored a little bit and tested it briefly under 4.1 between two cups of coffee.

**Comment by evadogstar** [comment]
@fepitre Is the code on the link still based on 0.7 version? Or you add everything from next version that I provide above or you only get maim args? By everything in new version I mean 1) imgurl logs 2) link to delete uploaded pics to each url in log 3) some bug fixes
It's a lot of improvements + now if it's not my last version I can not maintain it

**Comment by fepitre** [comment]
@evadogstar version is bumped after merge of new features/fixes. Current PR is only for fixing `scrot/maim` handling. We are going to bump to something like 0.7.1. When you will have done your new features/major changes we will bump it to probably 1.X. Don't worry about version, you can add features etc in separate branches and we take care of version. Also I would recommend you to sync your master branch with the `QubesOS-contrib` and then create new features/fixes from it. It helps to merge your work directly. If you need some help with `git` stuff you can ask here. You can also have a look at https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/proposing-changes-to-your-work-with-pull-requests.

**Comment by evadogstar** [comment]
@fepitre Yes, I think about the same to sync my current version from contrib version, but now conrtib version lose described above features. So after sync I will lose them and need to start again from integration. Ok, no problem. Time will show what to do later. 
p.s. imgurl log and opportunity to delete screenshots is the deal, btw

**Comment by evadogstar** [comment]
new version for Qubes R4.1 based on deepin-screenshot
https://github.com/evadogstar/qvm-screenshot-tool
much easy and intuitive way to draw, annotate screenshots. 
scrot no longer need

**Comment by Che15ea** [comment]
Hi
Most of my friends complain and wish to give up at the point of experiencing the complication of screenshot function while it can be much easier without compromising any security. I think it helps significantly to implement crop function when the screenshot is made, and a GUI dialog box to send to a selected VM or a default VM.
Despite that I use alias script to simplify the process. But it still takes many unnecessary steps to crop and send a picture: screenshot - use terminal to move to vm - crop with GIMP - send to another vm - share

**Comment by andrewdavidwong** [comment]
If it's already available, should this issue be closed?

**Comment by Che15ea** [comment]
Hi @tlaurion @andrewdavidwong 
Thank you for your reply!
Is it feasible to make it more newbie friendly and fool proof? Like pressing "PrintScreen" key and popping up crop and send-to-vm function by default without the need of installing anything or running any code.
I am a geek and I am OK with current implementation but it is so difficult for my journalist and activist friends to adapt the current Screenshot function. I don't think adding command in keyboard shortcut or manually installing the is newbie friendly. It needs lots of research to implement the qvm-screenshot-tool.
Sincerely

**Comment by Che15ea** [comment]
> @andrewdavidwong
> 
> From OP "The ability send dom0 screenshots directly to an AppVM/DispVM from the screenshot app is an oft-requested feature (see below). The ability to transfer saved screenshot files from dom0 to other VMs is already available, but it is neither obvious nor easy for most users to do this."
> 
> It is still geeky to deploy, is not deployed as default and requires users to customize things.
> 
> As @Che15ea just replied, I do not think original goals of opened ticket are met. If that is a goal, then the default printscreen and at-printscreen shortcuts should be already customized and qvm-screenshot-tool deployed by default under Q4.2?

I think so, crop and send to vm is a function that everyone needs and all my journalist and activist friends start to be frustrated about Qubes at this point. It should not take everybody more than 20 minutes to setup.

**Comment by DemiMarie** [comment]
@Che15ea I agree with this.  I think this should be installed by default.  @marmarta what do you think?

**Comment by unman** [comment]
Once again, different experiences.
The people I support don't have issues with copying files out of dom0 ,
regardless of their level of experience.
They tend NOT to use this package, and just use a screen shot tool in
dom0.

**Comment by lorenzog** [comment]
Cross-posting here (just noticed this thread, apologies for the duplicate) from https://github.com/QubesOS/qubes-issues/issues/5809#issuecomment-1286192955

TL;DR

Save this in dom0 and make executable:
```
#!/bin/sh
# find current window
WINDOW_ID=$(xprop -root _NET_ACTIVE_WINDOW | cut -d ' ' -f 5 | tr -d ',')
TARGET_VM=$(xprop -id $WINDOW_ID _QUBES_VMNAME | cut -d ' ' -f 3 | tr -d ''')
# save to clipboard, using rectangular selection
xfce4-screenshotter -c -r
# send to TARGET
xclip -out -sel c -t image/png | qvm-run --pass-io $TARGET_VM 'xclip -sel c -t image/png'
```
Go to settings and map the 'print screen' key to the script above. 
This will send the screenshot to the VM in focus when pressing 'print screen'. 
I'm still looking for a solution to have a nice dialog to ask the user which VM they want to send the screenshot to - something like the dialog for 'qvm-copy'.

**Comment by evadogstar** [comment]
@DemiMarie 
@unman People want tool to crop screenshots, annotate and upload to imgurl from the day one  :-) Sending screenshots to othe VM is only little part of issue.
@lorenzog tag me please when you will find how qvm-copydo this

> I think so, crop and send to vm is a function that everyone needs and all my journalist and activist friends start to be frustrated about Qubes at this point. It should not take every user's 20 minutes to setup.

@Che15ea Happy that my tool can help you. From day one on Qubes many years ago I experienced something the same, but it was not possible to take 20 minutes to setup and only one solution was to write it myself :-D :-D

**Comment by UndeadDevel** [comment]
For anyone who is having trouble getting @lorenzog 's code to work (e.g. xclip not installed by default on Qubes 4.1 dom0 and the method of grabbing focused VM throws syntax error for me), here is a different version:

```
#!/bin/sh
# lets the user take a screenshot based on rectangular selection and sends it to the currently focused VM

CUR_WIN_ID=`xdotool getwindowfocus`
CUR_VM=`xprop _QUBES_VMNAME -id $CUR_WIN_ID | cut -d \" -f 2`

if [[ "$CUR_VM" != "_QUBES_VMNAME:  not found." ]]; then
	xfce4-screenshooter -r -o "qvm-copy-to-vm $CUR_VM"
	notify-send "Screenshot sent!" "Your selection has been sent as a screenshot to $CUR_VM!"
fi
```

**Comment by lorenzog** [comment]
> For anyone who is having trouble getting @lorenzog 's code to work (e.g. xclip not installed by default on Qubes 4.1 dom0 and the method of grabbing focused VM throws syntax error for me), here is a different version:

Hi! Thanks for the update - in fact, as you noticed xclip stopped working since the ability to copy images with mime-type image/png disappeared from the latest version. There's a deep [rabbit hole](https://github.com/astrand/xclip/issues/38) that I haven't got into yet fully and probably never will. 

I can't seem to get your script to work though - mind you, I'm using i3 and not xfce at the moment, but the `-o` option doesn't seem to pass the output to the target VM.

**Comment by lorenzog** [comment]
Ok, I went down the rabbit hole :) this script now works in Qubes 4.1. It requires `xclip` to be present on the destination qube. 

``` 
#!/bin/sh
WINDOW_ID=$(xprop -root _NET_ACTIVE_WINDOW | cut -d ' ' -f 5 | tr -d ',')
echo "Window ID: $WINDOW_ID"
TARGET_VM=$(xprop -id $WINDOW_ID _QUBES_VMNAME | cut -d ' ' -f 3 | tr -d '"' )
echo "Target VM: $TARGET_VM"
DEST=$(mktemp)
echo "Dest: $DEST"
xfce4-screenshooter -r -s "$DEST"
cat "$DEST" | qvm-run --pass-io $TARGET_VM -- xclip -sel c -t image/png -l 1
rm "$DEST"
```
What needed to change:
1. xclip was "waiting" for the receiver (qvm-run) to pick up the content of the clipboard with the appropriate "target", and because qvm-run never asked for a target, it was returning an error (and an empty clipboard)
2. The screenshot is now saved to a temporary location on disk (`mkstemp` is not available in dom0?)
3. The `-l 1` instructs xclip to only listen for one selection (e.g. your target application in the qube) before closing itself. This means that screenshots won't be available more than once, unless you run a clipboard manager like parcellite.

**Comment by UndeadDevel** [comment]
> I can't seem to get your script to work though - mind you, I'm using i3 and not xfce at the moment, but the -o option doesn't seem to pass the output to the target VM.

Works perfectly for me on a pretty fresh Q4.1 install and doesn't require installing any packages...
It would be nice if the much more feature-rich tool was included by default in the qubes repo, though.

**Comment by CumpsD** [comment]
> I'm still looking for a solution to have a nice dialog to ask the user which VM they want to send the screenshot to - something like the dialog for 'qvm-copy'.

This would be an awesome addition. Or the ability to save the screenshot in the global clipboard, so one can paste it in a VM of choice after grabbing the screenshot.

**Comment by CumpsD** [comment]
If it helps, I made this:

```bash
cat qvm-copy-file-to-clipboard 
#!/usr/bin/bash
set -e -o pipefail
#
# The Qubes OS Project, http://www.qubes-os.org
#
# Copyright (C) 2015  Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
#

if [ $# -lt 1 ] ; then
    echo usage: $0 'file'
    exit 1
fi

#echo "base64 \"$@\" > /var/run/qubes/qubes-clipboard.bin"
base64 "$@" > /var/run/qubes/qubes-clipboard.bin

#echo "echo -n dom0 > /var/run/qubes/qubes-clipboard.bin.source"
echo -n dom0 > /var/run/qubes/qubes-clipboard.bin.source

if [ "${0##*/}" = "qvm-copy-file-to-clipboard" ]; then
	rm -rf -- "$@"
fi
```

I then configured this as a custom action in the built in dom0 screenshot tool:

![image](https://github.com/QubesOS/qubes-issues/assets/1201644/b7eccc08-a484-419b-9ef5-a3558a32516f)

![image](https://github.com/QubesOS/qubes-issues/assets/1201644/f6e6a58c-c107-4b3b-955f-d22664399d3a)

After I take a screenshot, it copies the file as base64 in the inter-qube clipboard, which I can then CTRL+SHIFT+V into my target qube and run this to get the file in my clipboard to CTRL+V it into github/slack/....

```code
xclip -o | base64 -d | xclip -i -selection clipboard -t image/png
```

I need to come up with a nice way to trigger this last line in my target VM after pressing CTRL+SHIFT+V, maybe bind the script to yet another keybind?

**Comment by tlaurion** [comment]
Referred upstream to work done by @ben-grande for reuostreaming and the downstream under contrib at https://github.com/evadogstar/qvm-screenshot-tool/issues/10

**Comment by ben-grande** [comment]
> Referred upstream to work done by @ben-grande for reuostreaming and the downstream under contrib at [evadogstar/qvm-screenshot-tool#10](https://github.com/evadogstar/qvm-screenshot-tool/issues/10)
> 
> Discussion leading to implemtstiom changes prior of qusal inclusion happened at [ben-grande/qusal#22](https://github.com/ben-grande/qusal/issues/22)
> 
> This is a cleaner version of the script, passing through qvm-copy and removing dom0 file edition, limiting the scope to full-screen and active window screenshoting, also removing uploading option which is delegated to be done on the qube where those edit/uploading actions should be done IMO.

Changes can also be seen in the commit message https://github.com/ben-grande/qusal/commit/134a26a0f50104355d735649ee9cce1d0c875c3a.

> As current state, qvm-screenshot per qusal deploys it correctly and works out if the box through qusal. Unfortunately, setting shortcuts in KDE is not so straightforward as it is for xfce, and qusal deploys KDE per deployment of dom0 salt application.
> 
> In current state, this means switching from xfce/KDE when needs come to taking screenshot, which to be honest makes sense, since xfce is what is expected to be used by default. So screenshot helper for xfce showing full-screen UX is what people would expect in full-screen screenshot anyway.

You don't need to change from KDE to Xfce, you just need to configure the shortcuts manually for the time being.

> Now the question is how to upstream those changes so they make their way to contrib repo.
> 
> For commits and accountability I'm tagging @ben-grande here.
> 
> I guess the improvements of the scripts should land as a PR against https://github.com/QubesOS-contrib/qubes-qvm-screenshot-tool per contrib guidelines so this ticket can be closed and improvements issues can be opened.
> 
> qvm-screenshot is a contrib project under qubes-qvm-screenshot-tool, and contrib guidelines encourage PR against already existing contrib repositories.https://github.com/QubesOS/qubes-issues/issues/953#issuecomment-735701421

About contributing upstream, the repos [QubesOS-contrib/qubes-qvm-screenshot-tool](https://github.com/QubesOS-contrib/qubes-qvm-screenshot-tool) and [evadogstar/qvm-screenshot-tool](https://github.com/evadogstar/qvm-screenshot-tool) differ [since 2020](https://github.com/QubesOS/qubes-issues/issues/953#issuecomment-735701421). My version also removes a Imgurl, a tool which original author [@evadogstar finds important](https://github.com/QubesOS/qubes-issues/issues/953#issuecomment-1356769433). I don't know how many people it would affect if Imgurl was removed. Although all projects have the same goal of taking screenshot and moving to qubes, they differ in features. I did clean the code extensively more than add features.

I committing enhancements to QubesOS-Contrib is possible, but then it would differ from upstream evadogstar repo. I committing to evadogstar, but they don't [open PRs to QubesOS-Contrib](https://github.com/evadogstar/qvm-screenshot-tool/issues/8) for new changes. There is no clear path to upstream it.

**Comment by marmarek** [comment]
Since https://github.com/QubesOS/qubes-issues/issues/7739 is implemented, we have a way to define per-package maintainers now. So, I have a proposal to improve the screenshoting situation:
1. Move https://github.com/QubesOS-contrib/qubes-qvm-screenshot-tool to the main organization (and also to the main repository)
2. Add @ben-grande and/or @evadogstar (if you agree) as the maintainer of the package.
3. Configure builder to require 2 signed tags on this package - with the idea primary maintainer cares for the functionality and general code shape, and the other person is only verifying if no backdoors or obvious security issues are added (such checking of incremental changes shouldn't be much work given the size of this tool). The second tag can be from the other maintainer, or from qubes core team (for example @fepitre or @DemiMarie or me or ...). To be clear, the general direction and code shape of the tool would be up to the primary maintainer, I do not want nitpicking like "this could be done a bit faster another way" blocking that second tag.

There will need to be a bit coordination needed for which we don't have tooling yet, but I think we can manage even manually for now. Specifically: how to coordinate between those 2 people pushing tags. Ideally, we'd have some system where a PR(?) is merged only when two tags are prepared to be pushed - and then PR merge and two tags are pushed to the repo at the same time. But, in the meantime, we can simply have few people with push access to the repo, and have them coordinate via comments/mail/IM etc. This means at some times (before the second tag is pushed) builder will refuse to fetch the sources, and maybe at times some revert will be needed (when second reviewer spot some issue), but that's IMO still good enough for now.

What do you think about this proposal?

**Comment by ben-grande** [comment]
> 1. Move https://github.com/QubesOS-contrib/qubes-qvm-screenshot-tool to the main organization (and also to the main repository)

I also think it makes sense. It is just some lines of code and leaving it on QubesOS-Contrib make people wait for the original author to make changes while transferring it to the main organization is easier to know where to contribute to (directly).

> 2. Add @ben-grande and/or @evadogstar (if you agree) as the maintainer of the package.

I am not sure if *if you agree* was intended for both people or just the name close to the phrase, therefore I'm answering yes.

> 3. Configure builder to require 2 signed tags on this package - with the idea primary maintainer cares for the functionality and general code shape, and the other person is only verifying if no backdoors or obvious security issues are added (such checking of incremental changes shouldn't be much work given the size of this tool). The second tag can be from the other maintainer, or from qubes core team (for example @fepitre or @DemiMarie or me or ...). To be clear, the general direction and code shape of the tool would be up to the primary maintainer, I do not want nitpicking like "this could be done a bit faster another way" blocking that second tag.

Multiple taggers that have at least someone not from ITL would be a nice experiment. I don't want to expand this too much on the screenshot related thread, but lets consider there are at least 3 authorized keys, requiring 2 keys to be valid and 2 of the keys are not used by ITL: me, @fepitre and @evadogstar. Considering ITL keys to be always trustworthy, at least 1 signature needs to come from ITL, else 2 non-ITL keys could sign a malicious commit.

@marmarek suggested this on #7739.

> I don't exclude more complex schemes (3 tags from any of A,B,C,D,E or 2 tags if at least one is made by X), but lets start with a simpler option first.

Which doesn't seem to have been completed as far as I know: *if at least one is made by X*.

> There will need to be a bit coordination needed for which we don't have tooling yet, but I think we can manage even manually for now. Specifically: how to coordinate between those 2 people pushing tags. Ideally, we'd have some system where a PR(?) is merged only when two tags are prepared to be pushed - and then PR merge and two tags are pushed to the repo at the same time. But, in the meantime, we can simply have few people with push access to the repo, and have them coordinate via comments/mail/IM etc. This means at some times (before the second tag is pushed) builder will refuse to fetch the sources, and maybe at times some revert will be needed (when second reviewer spot some issue), but that's IMO still good enough for now.
> 
> What do you think about this proposal?

Coordination via comments on issues and PR is my preferred method of contact if I become maintainer.

**Comment by ben-grande** [comment]
Ping.

**Comment by ben-grande** [comment]
Maintainer @evadogstar has account deleted, the package is "orphaned".

The repo [QubesOS-contrib/qubes-qvm-screenshot-tool](https://github.com/QubesOS-contrib/qubes-qvm-screenshot-tool) has lost it's upstream and community maintainer.

The repo [evadogstar/qvm-screenshot-tool](https://github.com/evadogstar/qvm-screenshot-tool) had some issues opened by the Qubes Core Team at the time, such as @fepitre (to sync commits with Qubes) and @ninavizz (UX), which could be useful improvements ( I don't remember fully their contents)..... how to get that back now that upstream has deleted it's account?

---

Now on the policy side of opening issues to community owned repos, those things can happen (deletion)... if they were open in qubes-issues, at least a proper backup could have been made.

**Comment by fepitre** [comment]
I think we can  just either drop this package or consider it as the upstream. I don't think we specially need to move it to QubesOS directly. I'm fine to add you as maintainer and we would need to have at least me to add another tag. This is similar to https://github.com/QubesOS-contrib/qubes-app-split-browser where I'm reviewing most of the work done in PR and the maintainer is already preparing everything, I just need to add a tag on it for triggering the build process.

**Comment by ben-grande** [comment]
Fork made: https://github.com/ben-grande/qubes-qvm-screenshot-tool

**Comment by fepitre** [comment]
> Fork made: https://github.com/ben-grande/qubes-qvm-screenshot-tool

Please tell me when you have tested it with builderv2 in order to add it to the CI too.

**Comment by ben-grande** [comment]
Build with qubes-builderv2 tested and succeeding.

This is the config I used to build, I understand that `key-dirs`, `maintainers` and other key related options will be changed to your key and to use tags instead of signed commits.

```yaml
---

git:
  baseurl: https://github.com
  prefix: ben-grande/
  branch: main
  maintainers:
    - DF3834875B65758713D92E91A475969DE4E371E3

key-dirs:
  - ../qusal-builder/keys/
backend-vmm: xen
debug: true
verbose: true
qubes-release: r4.2
timeout: 3600

skip-git-fetch: false
fetch-versions-only: false

distributions:
  - host-fc37

components:
  - builder-rpm:
      branch: main
      packages: false
      url: https://github.com/QubesOS/qubes-builder-rpm
      maintainers:
        - 0064428F455451B3EBE78A7F063938BA42CFA724
  - qubes-qvm-screenshot-tool:
      branch: master
      verification-mode: less-secure-signed-commits-sufficient
      maintainers:
        - DF3834875B65758713D92E91A475969DE4E371E3
```

**Comment by ben-grande** [comment]
@fepitre ^

**Comment by ben-grande** [comment]
@marmarta can you please take a look at https://github.com/ben-grande/qubes-qvm-screenshot-tool and evaluate usability?

**Comment by marmarta** [comment]
> [@marmarta](https://github.com/marmarta) can you please take a look at https://github.com/ben-grande/qubes-qvm-screenshot-tool and evaluate usability?

Not great.

I used it as-is, with zenity, and when I selected the "window" option some screenshot was taken, but I don't know of what. I want a way to click on which window I'm screenshotting.

When you pick what to do with the screenshot, there shouldn't be three non-mutually exhaustive options to pick and a OK/Cancel buttons, this is super confusing, and also I have no idea what the options actually do (open file manager in qube? which qube?): what would make more sense was for there to be:
- a miniature of the screenshot, so that the user knows what screenshot was taken and if they clicked what they wanted to click
- clearly labeled, separate actions that do a thing without having to go through another screen; I think what would make sense is actions of "leave the screenshot in <name of guivm>", "move to another vm and open file manager there" (with the dropdown to select vm already there, "remove screenshot"; ideally those would be some sort of buttons you can just click and do the thing, but if not, definitely not having another popup window with copy to vm
- the select qube dropdown should work like the one in qrexec policy window, where you can start typing and see matching vms appear instantly (I could type in your separate window, but nothing happened?)

In general, this tool cannot be less useable than the existing xfce screenshot tool. I think its usability is a good starting point (also there are at worst two screens there, if you pick it from the menu), and if you use a keyboard shortcut there is just one, and that's the right number of screens.

**Comment by ben-grande** [comment]
> > [@marmarta](https://github.com/marmarta) can you please take a look at https://github.com/ben-grande/qubes-qvm-screenshot-tool and evaluate usability?
> 
> Not great.
> 
> I used it as-is, with zenity, and when I selected the "window" option some screenshot was taken, but I don't know of what. I want a way to click on which window I'm screenshotting.

It is the window you clicked right after. Can you run with xtrace so I know which screenshot tool was selected?

`sh -x ./qvm-screenshot`

> When you pick what to do with the screenshot, there shouldn't be three non-mutually exhaustive options to pick and a OK/Cancel buttons, this is super confusing,

The three options are:

- Exit (to quit the dialog)
- Open file manager in qube (opens the file manager in the target qube, which is on the next dialog)
- Move file (will be removed from the GUI domain after it is copied to the destination)

Multiple options can be selected, but it only makes sense to select the second and third option simulateneously, as clicking on the first "Exit", will stop it right there.

> and also I have no idea what the options actually do (open file manager in qube? which qube?): what would make more sense was for there to be:
> 
>     * a miniature of the screenshot, so that the user knows what screenshot was taken and if they clicked what they wanted to click

Ok, this can be done similarly to the monitor selection when using qubes video companionwith screenshare. But, this would change a lot of things, it would be easier to do such thing with python frameworks than simple shell tools.

>     * clearly labeled, separate actions that do a thing without having to go through another screen; I think what would make sense is actions of "leave the screenshot in ", "move to another vm and open file manager there" (with the dropdown to select vm already there, "remove screenshot"; ideally those would be some sort of buttons you can just click and do the thing, but if not, definitely not having another popup window with copy to vm

The last popoup is the qube selection. I understand you don't want a new popup, but this can be done with zenity or kdialog as far as I know.

>     * the select qube dropdown should work like the one in qrexec policy window, where you can start typing and see matching vms appear instantly (I could type in your separate window, but nothing happened?)

That's a bug, I don't remember this happening before, but it is definitely happening with zenity and working with kdialog.
 
> In general, this tool cannot be less useable than the existing xfce screenshot tool. I think its usability is a good starting point (also there are at worst two screens there, if you pick it from the menu), and if you use a keyboard shortcut there is just one, and that's the right number of screens.

To get one popup (qube selection):

```
qvm-screenshot --region --move
```

Yes, not nice as you either have to select `--move` or `--qube-file-manager` to skip the middle prompt. This middle prompt is not used by xfce4-screenshooter, I just thought it would be nice to have, but I can add an option to skip this prompt.

**Comment by marmarta** [comment]
Well, based on what you are saying, zenity / kdialog are just not good tools for this. 

> It is the window you clicked right after. Can you run with xtrace so I know which screenshot tool was selected?

xfce4-screenshot. I did not get to click on a window after selecting ok (also this tool  by itself has an "active window", not "window you clicked on" option, so that's probably that).



> The three options are:
> 
>     * Exit (to quit the dialog)
> 
>     * Open file manager in qube (opens the file manager in the target qube, which is on the next dialog)
> 
>     * Move file (will be removed from the GUI domain after it is copied to the destination)
> 
> 
> Multiple options can be selected, but it only makes sense to select the second and third option simulateneously, as clicking on the first "Exit", will stop it right there.

The fact that you have to explain this means this is not usable. Also if two options don't make sense together, they should not be selectable together. This is not a good design, I wrote up what would a better design be, and, as above, probably not doable with zenity/kdialog, which means those are not good tools for it.

Generally, if you want a decent interface for screenshots, xfce4-screenshot has one, it just lacks the option we need (copy/move to another qube).

**Comment by ben-grande** [comment]
I agree with Marta's point, but as I am without time, removing my assignment. Fixing the prompt messages and behavior is easy, but doing the GTK app would take time that I don't have. If anyone plans to do it, this PR has code to show window screenshot in GTK: https://github.com/QubesOS/qubes-video-companion/pull/26. If you adapt it a bit, you can show any image, in this case, the image already taken by the screenshot tool.

**Comment by tlaurion** [comment]
> I agree with Marta's point, but as I am without time, removing my assignment. Fixing the prompt messages and behavior is easy, but doing the GTK app would take time that I don't have. If anyone plans to do it, this PR has code to show window screenshot in GTK: https://github.com/QubesOS/qubes-video-companion/pull/26. If you adapt it a bit, you can show any image, in this case, the image already taken by the screenshot tool.

Please reconsider taking @ben-grande better code then upstream and actual screenshot tool. It's better and can be improved upon. Actual tool is a disaster

**Comment by marmarta** [comment]
> Please reconsider taking [@ben-grande](https://github.com/ben-grande) better code then upstream and actual screenshot tool. It's better and can be improved upon. Actual tool is a disaster

I don't think it is a good tool... it is worse than the default one UX-wise in my opinion.

**Comment by crass** [comment]
> Generally, if you want a decent interface for screenshots, xfce4-screenshot has one, it just lacks the option we need (copy/move to another qube).

This is now easily solved with xfce4-screenshot 1.10.0+, which R4.3 satisfies and probably R4.2, by using a custom action, as @CumpsD [uses in his solution above](https://github.com/QubesOS/qubes-issues/issues/953#issuecomment-1881967616).

> When you pick what to do with the screenshot, there shouldn't be three non-mutually exhaustive options to pick and a OK/Cancel buttons, this is super confusing, and also I have no idea what the options actually do (open file manager in qube? which qube?): what would make more sense was for there to be:
> 
> * a miniature of the screenshot, so that the user knows what screenshot was taken and if they clicked what they wanted to click

In R4.3, xfce4-screenshot already does this. The preview is a little small for me, but I can still get an idea of what was captured.

> * clearly labeled, separate actions that do a thing without having to go through another screen; I think what would make sense is actions of "leave the screenshot in ", "move to another vm and open file manager there" (with the dropdown to select vm already there, "remove screenshot"; ideally those would be some sort of buttons you can just click and do the thing, but if not, definitely not having another popup window with copy to vm

The custom actions of xfce4-screenshot can show the action as clearly labeled and selected. However, because those actions call out to a script. Having an additional screen to select VM is the path of least resistance. To have it integrated into xfce4-screenshot, we would have to carry a fork with those modifications, which I don't think is an option.

> * the select qube dropdown should work like the one in qrexec policy window, where you can start typing and see matching vms appear instantly (I could type in your separate window, but nothing happened?)

It would be nice if this VM selector widget that is already implemented for the qrexec policy window were more generally available. Or is it?

So, in my opinion the solution that strikes the right balance between effort, usability, and simplicity is to have this script at `/var/lib/qubes/qvm-copy-to-vm-screenshot` ([suggested here](https://forum.qubes-os.org/t/screenshot-dysfunctional/28653/19)):
```
#!/bin/bash

destination_vm=$(zenity --entry --title="Destination VM" --text="What VM shall I send this screenshot to?")
qvm-copy-to-vm "$destination_vm" "$1"

if [ $? -eq 0 ]; then
	notify-send -u normal -i dialog-information "success!" "$1 delivered to $destination_vm"
else
	notify-send -u critical -i dialog-information "failed!" "$1 not delivered to $destination_vm"
fi
```

Then add the script as a custom action to xfce4-screenshot. The user can add custom actions manually through the xfce4-screenshot preferences dialog. To have this from install, I think file should be added to the `xfce4-settings-qubes` package at path `/etc/xdg/xfce4/xfconf/xfce-perchannel-xml/xfce4-screenshot.xml.qubes` that contains:
```
<?xml version="1.0" encoding="UTF-8"?>

<channel name="xfce4-screenshooter" version="1.0">
  <property name="actions" type="empty">
    <property name="actions" type="int" value="1"/>
    <property name="action-0" type="empty">
      <property name="name" type="string" value="Copy to VM"/>
      <property name="command" type="string" value="/var/lib/qubes/qvm-copy-to-vm-screenshot %f"/>
    </property>
  </property>
</channel>
```
 
Ultimately, I would want to replace the zenity call in `qvm-copy-to-vm-screenshot` with the qrexec vm selector widget. Or even better, `qvm-copy-to-vm` should have an option to present that selector, so getting the `destination_vm` is not needed.

@marmarek @marmarta What do you think about this proposal?

**Comment by crass** [comment]
I've improved my proposal above to use a modified version of the RPC confirmation window, so the combo box has all the bells and whistles Marta was wanting. As such, I'm withdrawing my proposal above in favor of [my pull request](https://github.com/QubesOS/qubes-desktop-linux-xfce4/pull/34).

**Comment by notmuchtohide** [comment]
I've made [something simple](https://gist.github.com/notmuchtohide/5f23d7fd8f6d01e701e069187d8b68d7) that works for me. This could be an idea for a temporary solution, if anyone wants to try it.

The script is a xfce4-screenshooter custom action. It doesn't have new GUI code, at the cost of a "disposable-in-middle". 

To be able to send screenshots from dom0 to any VM, I thought I could send them through a disposable. The screenshot would go from dom0 to the disposable and there I call `qvm-move` and just pick the VM I want, with the existing qrexec policy GUI.

**Comment by deeplow** [comment]
> To be able to send screenshots from dom0 to any VM, I thought I could send them through a disposable. The screenshot would go from dom0 to the disposable and there I call `qvm-move` and just pick the VM I want, with the existing qrexec policy GUI.

Incidentally, a collateral benefit of this approach is that it might have protected dom0 from [QSB-118](https://www.qubes-os.org/news/2026/08/29/qsb-118/).
